File: src/cli/commands/note.rs
use crate::api::TriliumClient;
use crate::cli::output::{print_note, print_notes, print_success, print_tree, print_warning};
use crate::config::Config;
use crate::error::Result;
use crate::models::{CreateNoteRequest, NoteTreeItem, UpdateNoteRequest, CreateBranchRequest};
use crate::cli::args::NoteCommands;
use std::io::{self, Read, Write, IsTerminal};
use std::path::{Path, PathBuf};
use std::pin::Pin;
use std::future::Future;

pub async fn handle(command: NoteCommands, config: &Config, output_format: &str) -> Result<()> {
    let client = TriliumClient::new(config)?;

    match command {
        NoteCommands::Create {
            title,
            content,
            note_type,
            parent,
            edit,
        } => {
            let parent_id = parent.unwrap_or_else(|| config.default_parent_id.clone());
            
            let final_content = if edit {
                edit_content(content.as_deref(), config)?
            } else if content.is_none() && !io::stdin().is_terminal() {
                // Read from stdin if no content provided and stdin is piped
                let mut buffer = String::new();
                io::stdin().read_to_string(&mut buffer)?;
                if buffer.is_empty() {
                    return Err(crate::error::TriliumError::ApiError("No content provided and stdin is empty".to_string()));
                }
                buffer
            } else {
                content.unwrap_or_default()
            };

            let request = CreateNoteRequest {
                parent_note_id: parent_id,
                title: title.clone(),
                note_type,
                content: final_content,
                note_position: None,
                prefix: None,
                is_expanded: None,
                is_protected: None,
            };

            let note = client.create_note(request).await?;
            print_success(&format!("Created note: {} ({})", note.title, note.note_id));
            print_note(&note, output_format, false);
            Ok(())
        }

        NoteCommands::Get { note_id, content } => {
            let mut note = client.get_note(&note_id).await?;
            
            if content {
                let note_content = client.get_note_content(&note_id).await?;
                note.content = Some(note_content);
            }

            print_note(&note, output_format, content);
            Ok(())
        }

        NoteCommands::Update {
            note_id,
            title,
            content,
            edit,
        } => {
            let final_content = if edit {
                let existing = client.get_note_content(&note_id).await?;
                Some(edit_content(Some(&existing), config)?)
            } else {
                content
            };

            if title.is_none() && final_content.is_none() {
                print_warning("No changes specified");
                return Ok(());
            }

            let request = UpdateNoteRequest {
                title,
                note_type: None,
                mime: None,
                content: final_content.clone(),
                is_protected: None,
            };

            let note = client.update_note(&note_id, request).await?;
            
            // Update content separately if provided
            if let Some(content) = final_content {
                client.update_note_content(&note_id, &content).await?;
            }

            print_success(&format!("Updated note: {} ({})", note.title, note.note_id));
            Ok(())
        }

        NoteCommands::Delete { note_id, force } => {
            if !force {
                print!("Are you sure you want to delete note {}? [y/N] ", note_id);
                io::stdout().flush()
                    .map_err(|e| crate::error::TriliumError::InputError(format!("Failed to flush stdout: {}", e)))?;
                let mut input = String::new();
                io::stdin().read_line(&mut input)
                    .map_err(|e| crate::error::TriliumError::InputError(format!("Failed to read user input: {}", e)))?;
                if !input.trim().eq_ignore_ascii_case("y") {
                    print_warning("Deletion cancelled");
                    return Ok(());
                }
            }

            client.delete_note(&note_id).await?;
            print_success(&format!("Deleted note: {}", note_id));
            Ok(())
        }

        NoteCommands::List { parent_id, tree, depth } => {
            let notes = client.get_child_notes(&parent_id).await?;
            
            if tree {
                let mut tree_items = Vec::new();
                for note in notes {
                    let mut item = NoteTreeItem::new(note, 0);
                    build_tree(&client, &mut item, depth, 1).await?;
                    tree_items.push(item);
                }
                print_tree(&tree_items, None);
            } else {
                print_notes(&notes, output_format);
            }
            Ok(())
        }

        NoteCommands::Export { note_id, output, format } => {
            let data = client.export_note(&note_id, &format).await?;
            
            let output_path = output.unwrap_or_else(|| {
                let extension = match format.as_str() {
                    "html" => "html",
                    "markdown" => "md",
                    "zip" => "zip",
                    _ => "txt",
                };
                PathBuf::from(format!("{}.{}", note_id, extension))
            });

            std::fs::write(&output_path, data)?;
            print_success(&format!("Exported note to: {}", output_path.display()));
            Ok(())
        }

        NoteCommands::Import { file, parent, format } => {
            let content = std::fs::read(&file)?;
            
            let actual_format = if format == "auto" {
                detect_format(&file)
            } else {
                format
            };

            let parent_id = parent.as_deref().unwrap_or("root");
            let note = client.import_note(parent_id, content, &actual_format).await?;
            print_success(&format!("Imported note: {} ({})", note.title, note.note_id));
            Ok(())
        }

        NoteCommands::Move { note_id, parent_id } => {
            // For moving notes, we need to create a new branch and delete the old one
            let note = client.get_note(&note_id).await?;
            let current_branches = client.get_note_branches(&note_id).await?;
            
            // Create new branch
            let request = CreateBranchRequest {
                note_id: note_id.clone(),
                parent_note_id: parent_id.clone(),
                note_position: None,
                prefix: None,
                is_expanded: None,
            };
            
            let _new_branch = client.create_branch(request).await?;
            print_success(&format!("Moved note {} to parent {}", note.title, parent_id));
            
            // Optionally delete old branches if specified by the user
            if !current_branches.is_empty() {
                print_warning(&format!("Note still has {} other branch(es). Use branch commands to manage them.", current_branches.len()));
            }
            
            Ok(())
        }

        NoteCommands::Clone { note_id, clone_type } => {
            let note = client.get_note(&note_id).await?;
            
            let clone_request = CreateNoteRequest {
                parent_note_id: "root".to_string(),
                title: format!("{} (clone)", note.title),
                note_type: note.note_type.clone(),
                content: note.content.unwrap_or_default(),
                note_position: None,
                prefix: None,
                is_expanded: None,
                is_protected: None,
            };
            
            let cloned_note = client.create_note(clone_request).await?;
            
            if clone_type == "deep" {
                // For deep clone, we would need to recursively clone children
                print_warning("Deep cloning not fully implemented - only note content was cloned");
            }
            
            print_success(&format!("Cloned note: {} ({})", cloned_note.title, cloned_note.note_id));
            Ok(())
        }
    }
}

fn build_tree<'a>(
    client: &'a TriliumClient,
    item: &'a mut NoteTreeItem,
    max_depth: usize,
    current_depth: usize,
) -> Pin<Box<dyn Future<Output = Result<()>> + 'a>> {
    Box::pin(async move {
        if current_depth >= max_depth {
            return Ok(());
        }

        let children = client.get_child_notes(&item.note.note_id).await?;
        for child in children {
            let mut child_item = NoteTreeItem::new(child, current_depth);
            build_tree(client, &mut child_item, max_depth, current_depth + 1).await?;
            item.children.push(child_item);
        }

        Ok(())
    })
}

fn edit_content(initial_content: Option<&str>, config: &Config) -> Result<String> {
    use tempfile::NamedTempFile;
    use std::process::Command;
    use std::path::Path;

    let mut temp_file = NamedTempFile::new()?;
    if let Some(content) = initial_content {
        temp_file.write_all(content.as_bytes())?;
    }
    temp_file.flush()?;

    let editor = config.editor.clone()
        .or_else(|| std::env::var("EDITOR").ok())
        .unwrap_or_else(|| "vi".to_string());

    // Validate and sanitize the editor command
    let validated_editor = validate_editor(&editor)?;

    let status = Command::new(&validated_editor.command)
        .args(&validated_editor.args)
        .arg(temp_file.path())
        .status()?;

    if !status.success() {
        return Err(crate::error::TriliumError::Cancelled);
    }

    let content = std::fs::read_to_string(temp_file.path())?;
    Ok(content)
}

pub struct ValidatedEditor {
    pub command: String,
    pub args: Vec<String>,
}

pub fn validate_editor(editor_string: &str) -> Result<ValidatedEditor> {
    use crate::error::TriliumError;
    
    // First check the entire string for dangerous characters before splitting
    if editor_string.contains(';') || editor_string.contains('&') || editor_string.contains('|') 
        || editor_string.contains('$') || editor_string.contains('`') || editor_string.contains('<') 
        || editor_string.contains('>') || editor_string.contains('\n') || editor_string.contains('\r')
        || editor_string.contains('\'') || editor_string.contains('"') {
        return Err(TriliumError::SecurityError(
            "Editor command contains potentially dangerous characters".to_string()
        ));
    }
    
    // Split editor command and arguments safely
    let parts: Vec<&str> = editor_string.split_whitespace().collect();
    if parts.is_empty() {
        return Err(TriliumError::SecurityError("Empty editor command".to_string()));
    }

    let command = parts[0];
    let args: Vec<String> = parts[1..].iter().map(|&s| s.to_string()).collect();

    // Whitelist of allowed editors (common secure editors)
    let allowed_editors = &[
        "vi", "vim", "nvim", "nano", "emacs", "code", "gedit", "kate", 
        "subl", "atom", "notepad", "micro", "joe", "pico", "ed"
    ];

    // Extract just the executable name from the path
    let exe_name = Path::new(command)
        .file_name()
        .and_then(|name| name.to_str())
        .unwrap_or(command);

    // Check if the editor is in the whitelist
    if !allowed_editors.contains(&exe_name) {
        return Err(TriliumError::SecurityError(
            format!("Editor '{}' is not in the allowed list. Allowed editors: {}", 
                    exe_name, allowed_editors.join(", "))
        ));
    }

    // Note: We already validated the entire string for dangerous characters above

    Ok(ValidatedEditor {
        command: command.to_string(),
        args,
    })
}

fn detect_format(path: &PathBuf) -> String {
    match path.extension().and_then(|s| s.to_str()) {
        Some("html") | Some("htm") => "html".to_string(),
        Some("md") | Some("markdown") => "markdown".to_string(),
        Some("zip") => "zip".to_string(),
        _ => "text".to_string(),
    }
}

File: src/cli/commands/note.ts
import { createReadStream, existsSync, writeFileSync } from 'fs';
import { resolve, extname, basename } from 'path';

import chalk from 'chalk';
import type { Command } from 'commander';

import { TriliumClient } from '../../api/client.js';
import { Config } from '../../config/index.js';
import { TriliumError } from '../../error.js';
import type { NoteType, ExportFormat } from '../../types/api.js';
import type { OutputFormat } from '../../types/common.js';
import { formatOutput, handleCliError, createTriliumClient } from '../../utils/cli.js';
import { createLogger } from '../../utils/logger.js';
import type {
  BaseCommandOptions,
  NoteCreateOptions,
  NoteGetOptions,
  NoteUpdateOptions,
  NoteDeleteOptions,
  NoteListOptions,
  NoteExportOptions,
  NoteImportOptions,
  NoteMoveOptions,
  NoteCloneOptions,
} from '../types.js';

/**
 * Set up note management commands
 */
export function setupNoteCommands(program: Command): void {
  const noteCommand = program
    .command('note')
    .description('Note operations');

  // Create note
  noteCommand
    .command('create')
    .description('Create a new note')
    .argument('<title>', 'note title')
    .option('-c, --content <content>', 'note content')
    .option('-t, --note-type <type>', 'note type (text, code, etc.)', 'text')
    .option('-p, --parent <id>', 'parent note ID')
    .option('-e, --edit', 'open in editor')
    .action(async (title: string, options: NoteCreateOptions) => {
      const logger = createLogger(options.verbose);
      
      try {
        const client = await createTriliumClient(options);
        
        let content = options.content || '';
        
        // Open editor if requested or no content provided
        if (options.edit || (!content && !process.stdin.isTTY)) {
          const { openNoteInExternalEditor } = await import('../../utils/editor.js');
          const editorResult = await openNoteInExternalEditor(
            content,
            options.noteType || 'text'
          );
          
          if (editorResult.cancelled) {
            logger.info('Note creation cancelled');
            return;
          }
          
          content = editorResult.content;
        }
        
        const noteData = {
          title,
          content,
          type: (options.noteType || 'text') as NoteType,
          parentNoteId: options.parent || 'root'
        };
        
        const result = await client.createNote(noteData);
        
        const output = formatOutput([result.note], options.output, [
          'noteId', 'title', 'type', 'parentNoteIds', 'dateCreated'
        ]);
        console.log(output);
        
        if (options.output === 'table') {
          logger.info(chalk.green(`Note created successfully: ${result.note.noteId}`));
        }
        
      } catch (error) {
        handleCliError(error, logger);
      }
    });

  // Get note
  noteCommand
    .command('get')
    .description('Get note by ID')
    .argument('<note-id>', 'note ID')
    .option('-c, --content', 'include content')
    .option('-e, --edit', 'open in external editor')
    .action(async (noteId: string, options: NoteGetOptions & { edit?: boolean }) => {
      const logger = createLogger(options.verbose);
      
      try {
        const client = await createTriliumClient(options);
        
        // If edit mode, open in external editor
        if (options.edit) {
          const note = await client.getNoteWithContent(noteId);
          const { openNoteInExternalEditor } = await import('../../utils/editor.js');
          const editorResult = await openNoteInExternalEditor(
            note.content || '',
            note.type
          );
          
          if (editorResult.cancelled) {
            logger.info('Edit cancelled - no changes saved');
            return;
          }
          
          if (editorResult.changed) {
            await client.updateNoteContent(noteId, editorResult.content);
            logger.info(chalk.green('Note updated successfully'));
          } else {
            logger.info('No changes made to note');
          }
          return;
        }
        
        const note = options.content ? 
          await client.getNoteWithOptionalContent(noteId, true) : 
          await client.getNote(noteId);
        
        const columns = [
          'noteId', 'title', 'type', 'mime', 'isProtected', 
          'dateCreated', 'dateModified'
        ];
        
        // Handle content if it's a NoteWithContent
        const noteWithContent = note as any;
        if (options.content && noteWithContent.content) {
          columns.push('contentLength');
          noteWithContent.contentLength = `${noteWithContent.content.length} chars`;
          
          // Show content preview for table output
          if (options.output === 'table' && noteWithContent.content.length > 200) {
            noteWithContent.contentPreview = noteWithContent.content.substring(0, 200) + '...';
            columns.push('contentPreview');
          } else if (options.output === 'table') {
            columns.push('content');
          }
        }
        
        const output = formatOutput([note], options.output, columns);
        console.log(output);
        
        // Show full content for JSON output or if specifically requested
        if (options.content && noteWithContent.content && options.output === 'json') {
          console.log(JSON.stringify({ ...note, content: noteWithContent.content }, null, 2));
        }
        
      } catch (error) {
        handleCliError(error, logger);
      }
    });

  // Edit note in external editor (dedicated command)
  noteCommand
    .command('edit')
    .description('Edit note in external editor')
    .argument('<note-id>', 'note ID')
    .option('--editor <editor>', 'specify editor to use (overrides $EDITOR)')
    .option('--no-convert', 'do not convert HTML to Markdown')
    .action(async (noteId: string, options: BaseCommandOptions & { editor?: string; convert?: boolean }) => {
      const logger = createLogger(options.verbose);
      
      try {
        const client = await createTriliumClient(options);
        const note = await client.getNoteWithContent(noteId);
        
        const { openNoteInExternalEditor } = await import('../../utils/editor.js');
        const editorResult = await openNoteInExternalEditor(
          note.content || '',
          note.type,
          {
            editor: options.editor,
            convertHtmlToMarkdown: options.convert !== false
          }
        );
        
        if (editorResult.cancelled) {
          logger.info('Edit cancelled - no changes saved');
          return;
        }
        
        if (editorResult.changed) {
          await client.updateNoteContent(noteId, editorResult.content);
          
          if (options.output === 'json') {
            console.log(JSON.stringify({ success: true, noteId, updated: true }, null, 2));
          } else {
            logger.info(chalk.green(`Note ${noteId} updated successfully`));
          }
        } else {
          if (options.output === 'json') {
            console.log(JSON.stringify({ success: true, noteId, updated: false }, null, 2));
          } else {
            logger.info('No changes made to note');
          }
        }
      } catch (error) {
        handleCliError(error, logger);
      }
    });
  
  // Update note
  noteCommand
    .command('update')
    .description('Update existing note')
    .argument('<note-id>', 'note ID')
    .option('-t, --title <title>', 'new title')
    .option('-c, --content <content>', 'new content')
    .option('-e, --edit', 'open in editor')
    .action(async (noteId: string, options: NoteUpdateOptions) => {
      const logger = createLogger(options.verbose);
      
      try {
        const client = await createTriliumClient(options);
        
        const updates: any = {};
        
        if (options.title) updates.title = options.title;
        
        if (options.content) {
          updates.content = options.content;
        } else if (options.edit) {
          // Get current content for editing
          const currentNote = await client.getNoteWithContent(noteId);
          const { openNoteInExternalEditor } = await import('../../utils/editor.js');
          const editorResult = await openNoteInExternalEditor(
            currentNote.content || '',
            currentNote.type
          );
          
          if (editorResult.cancelled) {
            logger.info('Note update cancelled');
            return;
          }
          
          if (!editorResult.changed) {
            logger.info('No changes made to note');
            return;
          }
          
          updates.content = editorResult.content;
        }
        
        if (Object.keys(updates).length === 0) {
          throw new TriliumError('No updates specified. Use --title, --content, or --edit options.');
        }
        
        const note = await client.updateNote(noteId, updates);
        
        const output = formatOutput([note], options.output, [
          'noteId', 'title', 'type', 'dateModified'
        ]);
        console.log(output);
        
        if (options.output === 'table') {
          logger.info(chalk.green('Note updated successfully'));
        }
        
      } catch (error) {
        handleCliError(error, logger);
      }
    });

  // Delete note
  noteCommand
    .command('delete')
    .alias('rm')
    .description('Delete a note')
    .argument('<note-id>', 'note ID')
    .option('-f, --force', 'force delete without confirmation')
    .action(async (noteId: string, options: NoteDeleteOptions) => {
      const logger = createLogger(options.verbose);
      
      try {
        if (!options.force) {
          const readline = await import('readline');
          const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
          });

          const answer = await new Promise<string>((resolve) => {
            rl.question(
              chalk.yellow(`Are you sure you want to delete note ${noteId}? This action cannot be undone. (y/N): `),
              resolve
            );
          });
          rl.close();

          if (answer.toLowerCase() !== 'y' && answer.toLowerCase() !== 'yes') {
            logger.info('Delete cancelled.');
            return;
          }
        }

        const client = await createTriliumClient(options);
        await client.deleteNote(noteId);
        
        if (options.output === 'json') {
          console.log(JSON.stringify({ success: true, noteId }, null, 2));
        } else {
          logger.info(chalk.green(`Note ${noteId} deleted successfully`));
        }
        
      } catch (error) {
        handleCliError(error, logger);
      }
    });

  // List notes
  noteCommand
    .command('list')
    .alias('ls')
    .description('List child notes')
    .argument('[parent-id]', 'parent note ID (defaults to root)', 'root')
    .option('-t, --tree', 'show as tree')
    .option('-d, --depth <number>', 'maximum depth for tree view', (val) => parseInt(val, 10), 3)
    .action(async (parentId: string, options: NoteListOptions) => {
      const logger = createLogger(options.verbose);
      
      try {
        const client = await createTriliumClient(options);
        
        if (options.tree) {
          const tree = await client.getNoteTree(parentId, { depth: options.depth });
          
          if (options.output === 'json') {
            console.log(JSON.stringify(tree, null, 2));
          } else {
            displayTreeStructure(tree, 0, options.depth || 3);
          }
        } else {
          const notes = await client.getChildNotes(parentId);
          
          const output = formatOutput(notes, options.output, [
            'noteId', 'title', 'type', 'isProtected', 'dateModified'
          ]);
          console.log(output);
        }
        
        if (options.output === 'table') {
          logger.info(chalk.green(`Displayed notes for parent ${parentId}`));
        }
        
      } catch (error) {
        handleCliError(error, logger);
      }
    });

  // Export note
  noteCommand
    .command('export')
    .description('Export note')
    .argument('<note-id>', 'note ID')
    .option('-f, --format <format>', 'export format (html, markdown, pdf)', 'html')
    .option('-o, --output <file>', 'output file')
    .action(async (noteId: string, options: NoteExportOptions) => {
      const logger = createLogger(options.verbose);
      
      try {
        // Ensure output format is provided for createTriliumClient
        const clientOptions = { ...options, output: options.output || 'json' } as BaseCommandOptions;
        const client = await createTriliumClient(clientOptions);
        
        logger.info(`Exporting note ${noteId} as ${options.format}...`);
        
        const exportResult = await client.exportNote(noteId, (options.format || 'html') as ExportFormat);
        
        // Determine output path
        let outputPath = (options as any).output;
        if (!outputPath) {
          const note = await client.getNote(noteId);
          const extension = options.format === 'pdf' ? 'pdf' : 
                           options.format === 'markdown' ? 'md' : 'html';
          outputPath = `${note.title.replace(/[^a-zA-Z0-9]/g, '_')}.${extension}`;
        }
        
        // Write to file - exportResult is an ArrayBuffer
        const buffer = Buffer.from(exportResult);
        writeFileSync(outputPath, buffer);
        
        const formatData = [{
          noteId,
          format: options.format || 'html',
          outputFile: outputPath,
          size: `${buffer.byteLength} bytes`
        }];
        
        // Use the base options.output property, accessing through a type assertion
        const outputFormat: OutputFormat = (options as any).output || 'table';
        const output = formatOutput(formatData, outputFormat, ['noteId', 'format', 'outputFile', 'size']);
        console.log(output);
        
        if (outputFormat === 'table') {
          logger.info(chalk.green(`Note exported to ${outputPath}`));
        }
        
      } catch (error) {
        handleCliError(error, logger);
      }
    });

  // Import note
  noteCommand
    .command('import')
    .description('Import note from file')
    .argument('<file>', 'file to import')
    .option('-p, --parent <id>', 'parent note ID')
    .option('-f, --format <format>', 'import format (auto, html, markdown)', 'auto')
    .action(async (filePath: string, options: NoteImportOptions) => {
      const logger = createLogger(options.verbose);
      
      try {
        const resolvedPath = resolve(filePath);
        
        if (!existsSync(resolvedPath)) {
          throw new TriliumError(`File not found: ${resolvedPath}`);
        }
        
        const client = await createTriliumClient(options);
        
        // Auto-detect format if not specified
        let format = options.format;
        if (format === 'auto') {
          const ext = extname(resolvedPath).toLowerCase();
          if (ext === '.md') format = 'markdown';
          else if (ext === '.html' || ext === '.htm') format = 'html';
          else format = 'text';
        }
        
        logger.info(`Importing ${filePath} as ${format}...`);
        
        const { readFileSync } = await import('fs');
        const content = readFileSync(resolvedPath, 'utf8');
        const title = basename(resolvedPath, extname(resolvedPath));
        
        const result = await client.createNote({
          title,
          content,
          type: (format === 'markdown' ? 'text' : format) as NoteType,
          parentNoteId: options.parent || 'root'
        });
        
        const output = formatOutput([{
          noteId: result.note.noteId,
          title: result.note.title,
          importedFrom: filePath,
          format,
          size: `${content.length} chars`
        }], options.output, ['noteId', 'title', 'importedFrom', 'format', 'size']);
        console.log(output);
        
        if (options.output === 'table') {
          logger.info(chalk.green(`File imported as note: ${result.note.noteId}`));
        }
        
      } catch (error) {
        handleCliError(error, logger);
      }
    });

  // Move note
  noteCommand
    .command('move')
    .alias('mv')
    .description('Move note to another parent')
    .argument('<note-id>', 'note ID')
    .argument('<parent-id>', 'new parent note ID')
    .action(async (noteId: string, parentId: string, options: NoteMoveOptions) => {
      const logger = createLogger(options.verbose);
      
      try {
        const client = await createTriliumClient(options);
        
        const result = await client.moveNote(noteId, parentId);
        
        const output = formatOutput([{
          noteId,
          oldParent: 'unknown',
          newParent: parentId,
          moved: true
        }], options.output, ['noteId', 'oldParent', 'newParent', 'moved']);
        console.log(output);
        
        if (options.output === 'table') {
          logger.info(chalk.green(`Note ${noteId} moved to parent ${parentId}`));
        }
        
      } catch (error) {
        handleCliError(error, logger);
      }
    });

  // Clone note
  noteCommand
    .command('clone')
    .description('Clone note')
    .argument('<note-id>', 'note ID to clone')
    .option('-t, --clone-type <type>', 'clone type (deep, shallow)', 'deep')
    .action(async (noteId: string, options: NoteCloneOptions) => {
      const logger = createLogger(options.verbose);
      
      try {
        const client = await createTriliumClient(options);
        
        // For now, cloning creates another branch to root
        const result = await client.cloneNote(noteId, 'root');
        
        const output = formatOutput([{
          originalNoteId: noteId,
          branchId: result.branchId,
          parentNoteId: result.parentNoteId,
          cloneType: options.cloneType,
          success: true
        }], options.output, ['originalNoteId', 'branchId', 'parentNoteId', 'cloneType', 'success']);
        console.log(output);
        
        if (options.output === 'table') {
          logger.info(chalk.green(`Note cloned successfully with branch: ${result.branchId}`));
        }
        
      } catch (error) {
        handleCliError(error, logger);
      }
    });
}

/**
 * Display tree structure in console
 */
function displayTreeStructure(tree: any, depth: number = 0, maxDepth: number = 3): void {
  const indent = '  '.repeat(depth);
  const icon = tree.type === 'search' ? '🔍' : 
               tree.isProtected ? '🔒' : 
               tree.type === 'code' ? '💻' : '📄';
  
  console.log(`${indent}${icon} ${tree.title} (${tree.noteId})`);
  
  if (tree.children && depth < maxDepth) {
    tree.children.forEach((child: any) => {
      displayTreeStructure(child, depth + 1, maxDepth);
    });
  }
}